native/cocos/base/HasMemberFunction.h

```diff
@@ -0,0 +1,97 @@
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+
+#pragma once
+
+#include <type_traits>
+
+/* This macro is used for checking whether a member function exists in a class in compiling time.
+Usage:
+ If we have a class called `MyClass` as follows:
+
+ ```c++
+ class MyClass {
+ public:
+    int myMethod(bool a, float b) { return 100; }
+ };
+```
+
+Test code:
+```c++
+ #include "base/HasMemberFunction.h"
+
+ CC_DEFINE_HAS_MEMBER_FUNC(myMethod)
+
+ template <typename T>
+ void myTest(T* arg0) {
+    if constexpr (has_myMethod<T, int(bool, float)>::value) {
+        // <1>: DO SOMETHING if T owns `myMethod` function.
+        // ...
+    } else {
+        // <2>: DO SOMETHING if T doesn't own `myMethod` function.
+        // ...
+    }
+ }
+
+ static int myTestEntry() {
+    MyClass a;
+    myTest(&a); // --> Go to <1>
+    int b;
+    myTest(&b; // --> Go to <2>
+ }
+
+```
+*/
+
+#define CC_DEFINE_HAS_MEMBER_FUNC(memFunc)                                    \
+    template <typename, typename T>                                           \
+    struct has_##memFunc {                                                    \
+        static_assert(                                                        \
+            std::integral_constant<T, false>::value,                          \
+            "Second template parameter needs to be of function type.");       \
+    };                                                                        \
+                                                                              \
+    template <typename C, typename Ret, typename... Args>                     \
+    struct has_##memFunc<C, Ret(Args...)> {                                   \
+    private:                                                                  \
+        template <typename T>                                                 \
+        static constexpr auto check(T*)                                       \
+            -> typename std::is_same<                                         \
+                decltype(std::declval<T>().memFunc(std::declval<Args>()...)), \
+                Ret>::type;                                                   \
+                                                                              \
+        template <typename>                                                   \
+        static constexpr std::false_type check(...);                          \
+        typedef decltype(check<C>(0)) type;                                   \
+                                                                              \
+    public:                                                                   \
+        static constexpr bool value = type::value;                            \
+    };
+
+namespace cc {
+
+CC_DEFINE_HAS_MEMBER_FUNC(setScriptObject)
+CC_DEFINE_HAS_MEMBER_FUNC(getScriptObject)
```

Comment:
I think it is better to define it in conversion.h as it is only used there.

Reply:
Is there any other codes need them?

Reply:
It's also used in bindings/jswrapper/v8/Object.h now.
```c++
    template <typename T>
    inline void setPrivateObject(TypedPrivateObject<T> *data) {
        setPrivateObject(static_cast<PrivateObjectBase *>(data));
        if constexpr (cc::has_setScriptObject<T, void(Object *)>::value) {
            data->template get<T>()->setScriptObject(this);
        }
    }
```

Reply:
ok